#1 Implement application discovery service API client

tuanlc opened the pull request.

Commit:
#1 Bootstrap library

Commit:
#1 add linter

Commit:
#1 setup unit-test

Commit:
#1 implement application discovery service API client

Commit:
#1 add npm task to run both lint & unit tests

Commit:
#1 update documentation

rezk2ll commented on src/Client.js:
```diff
@@ -0,0 +1,20 @@
+  }
+
+  api(config) {
+    return this.httpClient(config).then((resp) => resp.data);
```
```suggestion
    return this.httpClient(config).then(({ data }) => data);
```

rezk2ll commented on test/.eslintrc.json:
```diff
@@ -0,0 +1,6 @@
+    "node": true,
+    "mocha": true
+  }
+}
```
```suggestion
}

```

itswillta commented on src/Client.js:
```diff
@@ -0,0 +1,17 @@
+const httpClient = require('./http-client');
```
Since this is a brand new library, can we use ES6 Modules' `import`, `export` syntax here?

tuanlc replied:
Yes, why not

itswillta commented on webpack.common.js:
```diff
@@ -0,0 +1,18 @@
+  output: {
+    path: path.resolve(__dirname, 'dist'),
+    library: 'ESNApiClient',
+    libraryTarget: 'umd'
```
And we can also output `esm` here together with `umd`.

tuanlc replied:
Can you elaborate? Are you mentioning to https://github.com/purtuga/esm-webpack-plugin?

itswillta replied:
Webpack needs a whole plugin to do that? :fearful:

I think we can create an issue, something like "Also output ESM", and handle it later. I think we'll need to use `rollup` to handle that.

itswillta replied:
`rollup` is just much much simpler for building libraries.

piavgh commented on webpack.prod.js:
```diff
@@ -0,0 +1,18 @@
+const common = require('./webpack.common.js');
+
+module.exports = merge(common, {
+  output: {
```
This is a library, why do we need to bundle it with Webpack?

tuanlc replied:
Firstly, can you tell me why not?

piavgh replied:
because when you require it in another code (for example, our SPA), it will be bundled using Webpack of that SPA.

Also, because we don't need to create a separated minified version to use in the browser (like <script src="esn-api-client.min.js></script>), so basically the Webpack script of this module will never be used.

piavgh replied:
just like module https://github.com/OpenPaaS-Suite/esn-frontend-videoconference-calendar, it is only required in other projects, so no Webpack needed

itswillta commented on src/Client.js:
```diff
@@ -0,0 +1,20 @@
+
+    this.httpClient = httpClient({ baseURL });
+
+    applicationDiscoveryService(this);
```
So this Client will be a big one that holds every single OpenPaaS API possible like this, right?

```js
    applicationDiscoveryService(this);
    calendarService(this);
    inboxService(this);
    ...
```

In my opinion, that might not be what we want for a frontend client, since using a single endpoint requires importing everything.

itswillta replied:
How about having an abstract class `Client` and many `ADSClient`, `InboxClient`, `CalendarClient`, etc. that extends it. And then we'll only import the client we want to use.

tuanlc replied:
AFAIK, this client is responsible only for core APIs. For particular APIs like calendar, inbox, etc, the frontend client will be implemented in the corresponding SPAs.

itswillta replied:
And that Client will have the methods to communicate with the equivalent endpoints.

tuanlc replied:
For more discussion detail: https://chat.linagora.com/openpaas/pl/633drbncb7y4mgqdnfw6err7fo

itswillta replied:
I was just bringing some examples. Even the core APIs will be divided into multiple separate groups, so it would eventually become:

```
coreService1(this);
coreService2(this);
coreService3(this);
...
```

itswillta replied:
And what if I just want to use ADS, but not the other core APIs.

tuanlc replied:
Agree with you

tuanlc replied:
I modified the code to able to load necessary APIs instead of all

itswillta commented on src/api/application-discovery-service.js:
```diff
@@ -0,0 +1,129 @@
+const BASE_PATH = '/ads';
+
+function mixin(client) {
+  client.ads = {
```
If I'm not wrong, there's one catch if we do it this way: We lose the benefit of IntelliSense. It would be much nicer if we could type `whateverClient.` and then the code editor could suggest every available method with the information about how to use it.

itswillta commented on package.json:
```diff
@@ -0,0 +1,43 @@
+  "name": "esn-api-client",
+  "version": "0.0.1",
+  "description": "JS library for OpenPaaS ESN APIs",
+  "main": "src/index.js",
```
If you use Webpack to bundle & build the library, this field should point to the main file in the `dist` folder.

itswillta commented on package.json:
```diff
@@ -0,0 +1,43 @@
+  "dependencies": {
+    "axios": "0.19.2"
+  }
+}
```
And you might also want to add a `files` field that includes the `dist` folder (possibly also the `src` folder): https://docs.npmjs.com/files/package.json#files.

Commit:
#1 include src/ and dist/ folder when the library is installed as a dependency

Commit:
#1 the main file should point to the bundled file

Commit:
#1 refactor to able to use es6 modules syntax

itswillta commented on src/index.js:
```diff
@@ -1,3 +1,4 @@
-};
+import Client from './Client';
+
+// eslint-disable-next-line import/prefer-default-export
```
Can you disable this rule globally in `.eslintrc`, as named exports will be used quite often?

Commit:
#1 make UMD build available on both browsers and Node.js

For more detail: https://webpack.js.org/configuration/output/\#outputglobalobject

Commit:
#1 improve Client to able to load necessary APIs instead of all
